src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1468,15 +1468,22 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
                     memberType = CreateErrorType();
                     member = null;
                     hasErrors = true;
+                    isLengthOrCount = false;
                 }
                 else
                 {
                     member = LookupMembersForPropertyPattern(inputType, expr, diagnostics, ref hasErrors);
                     memberType = member.Type;
+                    // If we're dealing with the member that makes the type countable, and the type is also indexable, then it will be assumed to always return a non-negative value
+                    isLengthOrCount = memberType.SpecialType == SpecialType.System_Int32 &&
+                                      member.Symbol is { Name: WellKnownMemberNames.LengthPropertyName or WellKnownMemberNames.CountPropertyName, ContainingType: TypeSymbol containingType } memberSymbol &&
+                                      (memberSymbol.Equals(((MethodSymbol?)GetWellKnownTypeMember(WellKnownMember.System_Array__get_Length, BindingDiagnosticBag.Discarded, syntax: node))?.AssociatedSymbol) ||
+                                       TryPerformPatternIndexerLookup(node, containingType, argIsIndex: true, indexerAccess: out _, patternSymbol: out _, out PropertySymbol? lengthProperty, BindingDiagnosticBag.Discarded) &&
+                                       memberSymbol.Equals(lengthProperty)); // If both Count and Length are present we want the one that makes this type countable.
```

Comment:
Will these ever not be reference equal? Can we just do a simple `(object)memberSymbol == lengthProperty` here? If not, we should probably be using some kind of compare option to specify how the types should compare.

Reply:
>Will these ever not be reference equal? 

How could we know this for a fact? I think it'll be safer to just use Equals.

>we should probably be using some kind of compare option

The return type is a primitive, so I don't think options would make any difference.

Reply:
> How could we know this for a fact? I think it'll be safer to just use Equals.

I haven't looked through the implementation of TryPerformPatternIndexerLookup, but I suspect it will always be either reference equal or not equal at all. Let's make sure we're being correct here.

> The return type is a primitive

We're not comparing return types, we're comparing member symbols. What if those symbols are contained in types that are generic over `nint` vs `IntPtr`: should those be equal? What about `string?` vs `string`? If we don't think that these should be the _same_ symbol, then we need to define our expectations for the rest of the parts of the symbol and containing symbol.

Reply:
> So this happens sometimes which is weird?

That's actually pretty concerning (though likely unrelated to your change). @AlekseyTs, are there any member exposure issues you're currently tracking?

> The return type is exactly int, we don't accept any type substitution here.

The return type isn't what I'm worried about. What I'm worried about is the `ContainingType` of the member symbol. Because these are member symbols, not return types, everything involved in the member needs to be equal, including the type that contains the Length property. That type can be generic over anything it wants. For now, I think we should like be using `Symbol.Equals(memberSymbol, lengthProperty, TypeCompareKind.ConsiderEverything)`.